### ISSUE TITLE
Support multi-reference visual prompt aggregation for YOLOE

### ISSUE THREAD

**Issue by zgh2022** [ultralytics/ultralytics#25635]
## Description

YOLOE visual prompting currently supports multiple boxes and classes from one reference image, but it cannot aggregate exemplars from multiple reference images into one reusable class embedding profile. The public path calls `YOLOEVPDetectPredictor.get_vpe()`, which explicitly requires a single image:

```python
assert len(self.dataset) == 1, "get_vpe only supports one image!"
```

The documented multiple-image example applies separate prompts to multiple target images. It does not combine several reference images before inference.

I would like to add a focused multi-reference path using the existing visual prompt owner and the prompt embedding profile API merged in #25572.

## Proposed Behavior

A possible extension of the existing API is:

```python
model.predict(
    "target.jpg",
    refer_image=["reference-1.jpg", "reference-2.jpg"],
    visual_prompts={
        "bboxes": [boxes_for_reference_1, boxes_for_reference_2],
        "cls": [class_ids_for_reference_1, class_ids_for_reference_2],
    },
    predictor=YOLOEVPSegPredictor,
)
```

Class IDs would be global across reference images and remain sequential from zero. Each reference image may contain any subset of the classes.

For each reference image, the existing `LoadVisualPrompt` and SAVPE path would produce one prototype per class present in that image. Prototypes would then be:

1. aligned to the global class IDs,
2. averaged per class across reference images,
3. L2-normalized,
4. applied together through the existing `set_classes()` path.

This matches the class-wise aggregation already used by `YOLOEDetectValidator.get_visual_pe()`. All resulting class embeddings would be used in one target-image forward pass.

## Reuse and Export

No new export arguments or runtime model inputs are proposed.

After aggregation, users could use the existing API from #25572:

```python
model.save_prompt_embeddings("multi-reference.npz")

model = YOLOE("yoloe-26s-seg.pt")
model.load_prompt_embeddings("multi-reference.npz")
model.export(format="rknn", name="rk3588", quantize=16)
```

The exported ONNX, TensorRT, OpenVINO, CoreML, LiteRT, or RKNN model would remain a normal single-image-input static model with the prompt embeddings fused into its classification head.

## Use Case

A single visual example often overfits to its position, viewpoint, lighting, and scene context. Aggregating several examples of the same class produces a more stable visual prompt. Allowing different reference images to contain different classes also makes it practical to build a reusable visual vocabulary when all classes do not occur together in one scene.

I validated this aggregation strategy externally with YOLOE-26 on RK3588S. A two-class profile generated from multiple annotated references ran both classes in one FP16 RKNN forward. The joint raw class scores matched the corresponding single-class runs, while one target-image inference required one NPU call rather than one call per class. FP16 was retained because the tested W8A8 path noticeably degraded small-target visual-prompt quality.

## Scope

The proposed change would be limited to:

- multi-reference visual embedding extraction and class alignment,
- backward-compatible single-reference behavior,
- focused tests for same-class averaging, disjoint/mixed classes, invalid mappings, and single-reference compatibility,
- NPZ round-trip and static export smoke coverage,
- YOLOE documentation.

It would not change Predictor or AutoBackend contracts for exported models, RKNN conversion, NPU scheduling, `default.yaml`, or the existing static export behavior.

Would this focused extension of `refer_image` be acceptable, or would maintainers prefer a separate public visual-embedding aggregation method that callers pass to `set_classes()`?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @zgh2022, thank you for your detailed feature proposal and continued interest in Ultralytics 🚀! This is an automated response, and an Ultralytics engineer will also assist soon.

For additional context, please review the [Docs](https://docs.ultralytics.com/) and existing [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples. If you are proposing a change to the current YOLOE visual prompting behavior, please include a focused [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example), along with:

- The exact Ultralytics version and commit used
- A small set of reference and target images, or a self-contained script that generates them
- The complete `refer_image`, `visual_prompts`, and predictor configuration
- Expected versus actual behavior
- Relevant logs, prototype shapes, class mappings, and exported model details
- Any test results from the proposed aggregation, NPZ round-trip, and static export workflows

For custom training or model behavior questions, please also provide dataset examples and logs and verify the [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips).

## Upgrade

Please upgrade to the latest `ultralytics` package, including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml), in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) before testing:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments, with dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/), and [PyTorch](https://pytorch.org/) preinstalled:

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI verifies correct operation across YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Please share the requested reproducible details so the team can evaluate the proposed API direction, compatibility requirements, and test coverage. For real-time discussion, join [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Visit [Discourse](https://community.ultralytics.com/), or join conversations on our [Subreddit](https://www.reddit.com/r/Ultralytics/).

**Comment by zgh2022** [comment]
I completed a focused prototype and the requested reproducible validation against current `upstream/main` commit `81d076f8c` (`ultralytics 8.4.115`). The review branch is available here without an upstream PR while this feature request awaits approval:

https://github.com/zgh2022/ultralytics/tree/feature/yoloe-multi-reference-prompts

## Reproduction

This uses only the two existing Ultralytics assets. Class 0 (`person`) appears in both references, while class 1 (`bus`) appears only in the first reference:

```python
import numpy as np

from ultralytics import YOLOE
from ultralytics.models.yolo.yoloe import YOLOEVPSegPredictor

model = YOLOE("yoloe-26n-seg.pt")
visual_prompts = {
    "bboxes": [
        np.array([[221.52, 405.8, 344.98, 857.54], [20, 220, 805, 760]], dtype=np.float32),
        np.array([[100, 170, 720, 715], [730, 30, 1275, 715]], dtype=np.float32),
    ],
    "cls": [np.array([0, 1]), np.array([0, 0])],
}
results = model.predict(
    "ultralytics/assets/bus.jpg",
    refer_image=["ultralytics/assets/bus.jpg", "ultralytics/assets/zidane.jpg"],
    visual_prompts=visual_prompts,
    predictor=YOLOEVPSegPredictor,
    imgsz=640,
    conf=0.05,
)
model.set_classes(["person", "bus"], model.model.pe)
model.save_prompt_embeddings("person-bus-multiref.npz")
```

Observed prototype and prediction results on an RTX 3090:

```text
embedding shape: (1, 2, 512)
embedding L2 norms: [[1.0, 1.0]]
detections: 6
class IDs: [0, 0, 0, 1, 0, 0]
scores: [0.9746, 0.9244, 0.9199, 0.9144, 0.4635, 0.1067]
```

The implementation remaps each reference's present global class IDs to local contiguous IDs for the existing SAVPE path, then restores the global slots, averages one prototype per image/class, and L2-normalizes the result. This matches `YOLOEDetectValidator.get_visual_pe()` weighting, so multiple boxes of one class in one image still form one spatial prompt.

## NPZ and static export

The official NPZ round-trip reproduced the aggregated tensor exactly (`torch.equal=True`). Loading that profile into the matching detection architecture and exporting ONNX produced a normal single-image-input model:

```python
model = YOLOE("yoloe-26n.yaml").load("yoloe-26n-seg.pt")
model.load_prompt_embeddings("person-bus-multiref.npz")
model.export(format="onnx", imgsz=640, simplify=True)
```

Both the configured PT detector and ONNX Runtime returned 6 detections containing both classes on `bus.jpg`.

The same profile was exported through the unmodified official FP16 RKNN path:

```python
model.export(format="rknn", name="rk3588", quantize=16, imgsz=640, simplify=True)
```

The output was a 7.0 MiB static RKNN directory with one image input and metadata names `{0: person, 1: bus}`. Standard Ultralytics directory inference was then run on an Orange Pi RK3588S using RKNNLite 2.3.0 and NPU driver 0.8.8. No NPZ was copied to or loaded by the board:

```text
detections: 5
class IDs: [0, 1, 0, 0, 0]
names: {0: person, 1: bus}
scores: [0.9312, 0.9282, 0.8579, 0.8354, 0.1597]
NPU inference: 59.03 ms
```

## Focused tests

The branch adds one mocked aggregation/NPZ regression test to the existing test file. It covers three references with overlapping and disjoint class subsets, same-class boxes, global/local class remapping, normalization, NPZ reload, and rejection of non-sequential global IDs. The existing single-reference verbose regression also passes:

```text
2 passed in 5.50s
Ruff format/check: passed
```

The implementation changes only the existing `YOLOE.predict()` owner plus one existing test and the YOLOE documentation. It adds no public helper, exporter option, backend input, or configuration key. I can adjust the API to a separate embedding method instead if that is preferred. Please confirm whether this scope and the `refer_image=[...]` extension are acceptable before I open the PR.

**Comment by Y-T-G** [comment]
Duplicate of https://github.com/ultralytics/ultralytics/issues/21479
